File: src/main/scala/uk/co/developmentanddinosaurs/gatling/lambda/action/InvokeActionBuilder.scala
package uk.co.developmentanddinosaurs.gatling.lambda.action

import io.gatling.core.action.Action
import io.gatling.core.action.builder.ActionBuilder
import io.gatling.core.structure.ScenarioContext
import io.gatling.core.util.NameGen
import uk.co.developmentanddinosaurs.gatling.lambda.protocol.LambdaProtocol
import uk.co.developmentanddinosaurs.gatling.lambda.request.LambdaAttributes

case class InvokeActionBuilder(attr: LambdaAttributes)
    extends ActionBuilder
    with NameGen {
  override def build(ctx: ScenarioContext, next: Action): Action = {
    val protocol = getProtocol(ctx)
    val client = protocol.lambdaClient
    new InvokeAction(client, ctx.coreComponents, next, genName(""), attr)
  }

  private def getProtocol(ctx: ScenarioContext) = {
    ctx.protocolComponentsRegistry
      .components(LambdaProtocol.lambdaProtocolKey)
      .lambdaProtocol
  }

}

File: src/main/java/uk/co/developmentanddinosaurs/gatling/lambda/javaapi/action/InvokeActionBuilder.java
package uk.co.developmentanddinosaurs.gatling.lambda.javaapi.action;

import io.gatling.javaapi.core.ActionBuilder;
import io.gatling.javaapi.core.internal.Expressions;

public class InvokeActionBuilder implements ActionBuilder {

  private final uk.co.developmentanddinosaurs.gatling.lambda.action.InvokeActionBuilder wrapped;

  public InvokeActionBuilder(
      uk.co.developmentanddinosaurs.gatling.lambda.action.InvokeActionBuilder wrapped) {
    this.wrapped = wrapped;
  }

  public InvokeActionBuilder payload(String payload) {
    return new InvokeActionBuilder(wrapped.payload(Expressions.toStringExpression(payload)));
  }

  public InvokeActionBuilder requestName(String requestName) {
    return new InvokeActionBuilder(
        wrapped.requestName(Expressions.toStringExpression(requestName)));
  }

  @Override
  public io.gatling.core.action.builder.ActionBuilder asScala() {
    return wrapped;
  }
}
